### DISCUSSION TITLE
Programmatically generate WorkflowDefinition

### DISCUSSION THREAD

**Discussion by hobozero** [elsa-workflows/elsa-core#4756] [General]
Is there any documentation or could someone provide any guidance whatsoever on creating a workflow definition from programmatically generated Activities? We can instantiate a IActivitySerializer and successfully serialize what would be at the root of a WorkflowDefinition, but the proper invocation of the WorkflowBuilder and associted classes to generate the whole definition is not apparent. For example, I'd assume it's not necessry to manually assign IDs to each activity to serialize the workflow:

```csharp
var writeLine1 = new WriteLine("Please tell me your name:")
{
    Id = Guid.NewGuid().ToString(),
};
var writeLine2 = new ReadLine(nameVariable)
{
    Id = Guid.NewGuid().ToString(),
};
var writeLine3 = new WriteLine(context => $"Nice to meet you, {nameVariable.Get(context)}!")
{
    Id = Guid.NewGuid().ToString(),
};

var flow = new Flowchart
{
    // Register the name variable.
    Variables = { nameVariable }, 
    
    // Add the activities.
    Activities =
    {
        writeLine1, 
        writeLine2,
        writeLine3
    },
    
    // Setup the connections between activities.
    Connections =
    {
        new Connection(writeLine1, writeLine2),
        new Connection(writeLine2, writeLine3)
    }
};
var serializer = serviceProvider.GetRequiredService<IActivitySerializer>();
var workflowJson = serializer.Serialize(flow);
Console.WriteLine(workflowJson);

```

The objective is to programmatically create a workflow definition model that may be displayed in the dashboard and/or sent to an isolated Elsa server for execution. I'm sure it's obvious, but with the documentation still as a work in progress, we're coming up empty.

I would be happy to contribute docs if we can get a nudge in the right direction. Thank you!

**Comment by hobozero** [comment]
[Today's commit](https://github.com/elsa-workflows/elsa-core/commit/fa21cba51f0e0dc0532df22e4558e3cfaad9dd9b) by @sfmskywalker seems to add support for dynamic workflows. Is this the best approach to map a workflow to a WorkflowDefinition for import in the Studio Dashboard or execution engine. 
```csharp
var workflowBuilderFactory = serviceProvider.GetRequiredService<IWorkflowBuilderFactory>();
var mapper = serviceProvider.GetRequiredService<WorkflowDefinitionMapper>();
var workflowRegistry = serviceProvider.GetRequiredService<Elsa.Workflows.Runtime.Contracts.IWorkflowRegistry>();
var workflowBuilder = workflowBuilderFactory.CreateBuilder();

var workflow = await workflowBuilder.BuildWorkflowAsync<ProvisionPlanWorkflow>();
//store in memory
await workflowRegistry.RegisterAsync(workflow);

//retrieve from memory
var memoryWorkflowDefinitionStore = serviceProvider.GetRequiredService<MemoryWorkflowDefinitionStore>();
var searchFilter = new WorkflowDefinitionFilter()
{
    Id = workflow.Identity.Id,
};
var workflowDefinition = await memoryWorkflowDefinitionStore.FindAsync(searchFilter);

var workflowDefinitionModel = await mapper.MapAsync(workflowDefinition);

var apiSerializer = serviceProvider.GetRequiredService<IApiSerializer>();
var workflowJson = apiSerializer.Serialize(workflowDefinitionModel);
```

**Reply by sfmskywalker** [reply]
If the goal is to simply get the workflow in the database, then all you would need is this:

```csharp
var workflowBuilderFactory = serviceProvider.GetRequiredService<IWorkflowBuilderFactory>();
var workflowRegistry = serviceProvider.GetRequiredService<Elsa.Workflows.Runtime.Contracts.IWorkflowRegistry>();
var workflowBuilder = workflowBuilderFactory.CreateBuilder();

var workflow = await workflowBuilder.BuildWorkflowAsync<ProvisionPlanWorkflow>();

// This will store the workflow in the database, after which you can also open it in Studio.
await workflowRegistry.RegisterAsync(workflow);
```

**Reply by hobozero** [reply]
The goal is to obtain the [JSON itself](https://github.com/elsa-workflows/elsa-core/files/13854417/sample-workflow.json), in-process, so it can be imported into the dashboard or other engine instances without transferring via a shared data store. 

Currently, we're using the new `IWorkflowRegistry` with a `MemoryWorkflowDefinitionStore` behind it to handle the Activity type registration, ID creation, and internal materialization of a `Workflow` to a `Elsa.Workflows.Management.Entities.WorkflowDefinition`. The only way we found to obtain the resulting `WorkflowDefinition` was directly from the `MemoryWorkflowDefinitionStore`, which feels magical since its registration happens automatically in the `WorkflowManagementFeature`. We'd thought about implementing our own `IWorkflowDefinitionStore`.

Either way it's then sent to the `WorkflowDefinitionMapper` to get an `Elsa.Workflows.Management.Models.WorkflowDefinitionModel`, which happens(?) to match the `Elsa.Api.Client.Resources.WorkflowDefinitions.Models.WorkflowDefinition`. It wasn't clear how a `WorkflowDefinitionModel` is derived for use over APIs. We're also using the `ApiSerializer` just in case. Feels like we're either missing something or are at high risk of instability.

We originally tried to programmatically emulate Elsa Studio, but it's not using `IWorkflowBuilder`. It looks like it maps the X6 graph directly to `Elsa.Api.Client.Resources.WorkflowDefinitions.Models.WorkflowDefinition` then to a `Elsa.Api.Client.Resources.WorkflowDefinitions.Models.WorkflowDefinitionModel` for serialization. Do I have that right?

Thanks @sfmskywalker or anyone else for helping confirm the direction before we get too heavily invested.

**Reply by hobozero** [reply]
The model translation looks like it's done in Refit maybe? Are `Elsa.Api.Client.Resources.WorkflowDefinitions.Models.WorkflowDefinition` and `Elsa.Api.Client.Resources.WorkflowDefinitions.Models.WorkflowDefinitionModel` kept in sync by convention, or is there an implementation we should leverage? Wrong question entirely?

**Reply by sfmskywalker** [reply]
I see. In that case, you don't need the `IWorkflowRegistry` at all - that's only necessary if you want to add a `Workflow` to the runtime layer at runtime. All you need to go from a `Workflow` object to JSON is to map the `Workflow` object to a `Elsa.Workflows.Management.Models.WorkflowDefinitionModel`. To simplify this, [I added a method](https://github.com/elsa-workflows/elsa-core/commit/c601076437ab5a04bc3cad5e0845d8f0649b7b5c) to the [IWorkflowSerializer](https://github.com/elsa-workflows/elsa-core/blob/main/src/modules/Elsa.Workflows.Management/Contracts/IWorkflowSerializer.cs#L15) service that allows you to do this:

```csharp
IWorkflowSerializer workflowSerializer; // Inject this service or resolve from service container otherwise.
var workflow = new Workflow { Root = new Sequence { Activities = { new WriteLine("Hello World!") } } };
var json = workflowSerializer.Serialize(workflow);
```

The resulting `json` string is the expected format you need to import it in Elsa Studio.
I added a [temporary sample endpoint](https://github.com/elsa-workflows/elsa-core/blob/main/src/bundles/Elsa.Server.Web/Endpoints/DynamicWorkflows/Post/Endpoint.cs) where you can see this in action.

> We originally tried to programmatically emulate Elsa Studio, but it's not using IWorkflowBuilder. It looks like it maps the X6 graph directly to Elsa.Api.Client.Resources.WorkflowDefinitions.Models.WorkflowDefinition then to a Elsa.Api.Client.Resources.WorkflowDefinitions.Models.WorkflowDefinitionModel for serialization. Do I have that right?

You got that spot-on.

> The model translation looks like it's done in Refit maybe? Are Elsa.Api.Client.Resources.WorkflowDefinitions.Models.WorkflowDefinition and Elsa.Api.Client.Resources.WorkflowDefinitions.Models.WorkflowDefinitionModel kept in sync by convention, or is there an implementation we should leverage? Wrong question entirely?

It's a good question. The reason we have two models that are almost identical is that for one scenario, we need the `Variables` property to be `ICollection<Variable>` and in another scenario we need it to be `ICollection<VariableDefinition>`. This is a bit of cruft that I am hoping to be able to spend some time on at some point to streamline into a single model.

**Reply by hobozero** [reply]
A Serialize method on `IWorkflowSerializer` is absolutely ideal- very clean and highly appreciated. I just successfully tested. Is there a target release date? 3.1.0?

**Comment by hobozero** [comment]
Hi @sfmskywalker. There's an issue with serialization. I can open a new issue if this isn't intended behavior. 

When the serializer invokes the `WorkflowDefinitionMapper` to map workflow variables, it uses a `VariableDefinitionMapper`, which does a .ToString() on the variable value resulting in values such as `{System.Dynamic.ExpandoObject}` for serializable but non-primitive types. We're attempting a workaround of wrapping the Expando Type to overload ToString() to perform serialization there, but I'm guessing that's a terrible idea, We'll probably revert to using primitives only that express their values via ToString().
